File: runme/mcgen/local/datan/exasrc/e4rn.c
#include <stdio.h>
#include <math.h>
#include "datsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__50 = 50;
static double c_d1 = 1., c_dp5 = .5, c_dm1 = -1.;

main(void)
{
    /* Initialized data */ /* Local variables */
    static integer i, j, k, m, nws, nt = 5000, mt = 500, nx = 40;
    static double xa[2] = { 0.,0. }, xb[2] = { 10.,10. }, x0 = -1., delx = .05,
                  rand[2], epar[2], dinv, hist[40], xtry[2], d,
                  v[100] /* was [2][50] */, w[2], x[100] /* was [2][50] */,
                  eperp[2], uipar, h1[2], h2[2], dd, dx[2], ui[2],
                  vi[2], vj[2], uj[2], delphi, uiperp, dsq, phi;
    /* System generated locals */
    integer i__1;
    double d__1, d__2;
/* identify program to user */
    printf("%s\n", "Program E4RN performs");
    printf("%s\n\n", "Monte Carlo simulation of molecular motion.");
/* initialization */
    d = 1.;
    dsq = d * d;
    delphi = .1396262222222222;
    nws = 1;
/* choose initial positions */
    for (i = 1; i <= 50; ++i) {
L10:
	rnecuy_(rand, &c__2);
	for (k = 1; k <= 2; ++k) {
	    xtry[k - 1] = xa[k - 1] + .5 + rand[k - 1] *
                          (xb[k - 1] - xa[k - 1] - d);
	}
	if (i > 1) {
/* repeat if there is overlap with existing particle */
	    i__1 = i - 1;
	    for (j = 1; j <= i__1; ++j) {
/* Computing 2nd power */
		d__1 = xtry[0] - x[(j << 1) - 2];
/* Computing 2nd power */
		d__2 = xtry[1] - x[(j << 1) - 1];
		if (d__1 * d__1 + d__2 * d__2 <= dsq) {
		    goto L10;
		}
	    }
	}
	x[(i << 1) - 2] = xtry[0];
	x[(i << 1) - 1] = xtry[1];
    }
/* set initial velocities */
    for (i = 2; i <= 50; ++i) {
	v[(i << 1) - 2] = 0.;
	v[(i << 1) - 1] = 0.;
    }
    rnecuy_(rand, &c__1);
    phi = rand[0] * 6.28318;
    v[0] = cos(phi) * 1.;
    v[1] = sin(phi) * 1.;
/* write header for output table */
    printf("%s\n", "The following lines are numerical");
    printf("%s\n", "histograms for the x-component of velocity");
    printf("%s\n", "velocity of the 50 particles in the simulation.");
    printf("%s\n", "Each line is a histogram for a fixed time");
    printf("%s\n", "Line number 1 shows that 49 particles are at");
    printf("%s\n", "rest. They are in the histogram bin around");
    printf("%s\n", "vx=0. One particle has nonzero velocity.");
/* loop over all times */
    i__1 = nt;
    for (m = 1; m <= i__1; ++m) {
	if (m % mt == 1) {
/* produce histogram */
	    smhsin_(hist, &x0, &delx, &nx);
	    for (j = 1; j <= 50; ++j) {
		smhsfl_(hist, &x0, &delx, &nx, &v[(j << 1) - 2], &c_d1);
	    }
            for (k = 1; k <= 40; ++k)
                printf("%2li", (integer)hist[k-1]);
            printf("\n");
	}
/* advance all particles */
	for (i = 1; i <= 50; ++i) {
	    for (k = 1; k <= 2; ++k) {
		x[k + (i << 1) - 3] += v[k + (i << 1) - 3] * .01;
	    }
	}
/* test for collisions */
	for (i = 1; i <= 50; ++i) {
/* test for collisions with wall */
	    for (k = 1; k <= 2; ++k) {
		if (x[k + (i << 1) - 3] <= xa[k - 1] + .5 ||
                    x[k + (i << 1) - 3] >= xb[k - 1] - .5) {
		    v[k + (i << 1) - 3] = -v[k + (i << 1) - 3];
		}
	    }
/* test for collisions with another particle */
	    if (i < 50) {
		for (j = i + 1; j <= 50; ++j) {
		    dx[0] = x[(i << 1) - 2] - x[(j << 1) - 2];
		    dx[1] = x[(i << 1) - 1] - x[(j << 1) - 1];
/* Computing 2nd power */
		    d__1 = dx[0];
/* Computing 2nd power */
		    d__2 = dx[1];
		    dd = d__1 * d__1 + d__2 * d__2;
		    if (dd <= dsq) {
			dinv = 1. / sqrt(dd);
			eperp[0] = dx[0] * dinv;
			eperp[1] = dx[1] * dinv;
			epar[0] = eperp[1];
			epar[1] = -eperp[0];
			mtxgcl_(v, vi, &c__2, &c__50, &i);
			mtxgcl_(v, vj, &c__2, &c__50, &j);
			mtxadv_(vi, vj, w, &c__2);
			mtxmsv_(w, w, &c_dp5, &c__2);
			mtxsbv_(vi, w, ui, &c__2);
			mtxdot_(ui, eperp, &uiperp, &c__2);
			mtxdot_(ui, epar, &uipar, &c__2);
			uiperp = -uiperp;
			mtxmsv_(eperp, h1, &uiperp, &c__2);
			mtxmsv_(epar, h2, &uipar, &c__2);
			mtxadv_(h1, h2, ui, &c__2);
			mtxadv_(ui, w, vi, &c__2);
			mtxmsv_(ui, uj, &c_dm1, &c__2);
			mtxadv_(uj, w, vj, &c__2);
			mtxpcl_(v, vi, &c__2, &c__50, &i);
			mtxpcl_(v, vj, &c__2, &c__50, &j);
		    }
		}
	    }
	}
    }
    return 0;
} /* main */

File: runme/mcgen/local/datan/exasrc/e4rn.f
      PROGRAM E4RN
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(N=50,R=.5D0,DT=1.D-2,V0=1.D0,PI=3.14159D0)
      DIMENSION XA(2),XB(2),X(2,N),DX(2),V(2,N),XTRY(2)
      DIMENSION VI(2),VJ(2),W(2),UI(2),UJ(2),H1(2),H2(2)
      DIMENSION EPERP(2),EPAR(2),RAND(2),HIST(40)
      DATA XA/0.D0,0.D0/,XB/10.D0,10.D0/,NT/5000/,MT/500/
      DATA NX/40/,X0/-1.D0/,DELX/.05D0/
C identify program to user
      WRITE(*,*)' Program E4RN performs'
      WRITE(*,*)' Monte Carlo simulation of molecular motion'
      WRITE(*,*)' '
C initialization
      D=2.D0*R
      DSQ=D*D
      DELPHI=2.D0*PI/45.D0
      NWS=1
C choose initial positions
      DO 40 I=1,N
10      CALL RNECUY(RAND,2)
        DO 20 K=1,2
          XTRY(K)=XA(K)+R+RAND(K)*(XB(K)-XA(K)-D)
20      CONTINUE
        IF(I.GT.1) THEN
C repeat if there is overlap with existing particle
          DO 30 J=1,I-1
            IF((XTRY(1)-X(1,J))**2+(XTRY(2)-X(2,J))**2.LE.DSQ)
     +      GO TO 10
30        CONTINUE
          END IF
          X(1,I)=XTRY(1)
          X(2,I)=XTRY(2)
40      CONTINUE
C set initial velocities
        DO 45 I=2,N
          V(1,I)=0.D0
          V(2,I)=0.D0
45      CONTINUE
        CALL RNECUY(RAND,1)
        PHI=2.D0*PI*RAND(1)
        V(1,1)=V0*COS(PHI)
        V(2,1)=V0*SIN(PHI)
C write header for output table
        WRITE(*,*)'The following lines are numerical'
        WRITE(*,*)'histograms for the x-component of velocity'
        WRITE(*,*)'velocity of the 50 particles in the simulation.'
        WRITE(*,*)'Each line is a histogram for a fixed time'
        WRITE(*,*)'Line number 1 shows that 49 particles are at'
        WRITE(*,*)'rest. They are in the histogram bin around'
        WRITE(*,*)'vx=0. One particle has nonzero velocity.'
C loop over all times
        DO 1000 M=1,NT
          IF(MOD(M,MT).EQ.1) THEN
C produce histogram
            CALL SMHSIN(HIST,X0,DELX,NX)
            DO 47 J=1,N
              CALL SMHSFL(HIST,X0,DELX,NX,V(1,J),1.D0)
47          CONTINUE
            WRITE(*,'(1X,40I2)')(INT(HIST(K)),K=1,40)
          END IF
C advance all particles
          DO 60 I=1,N
            DO 50 K=1,2
              X(K,I)=X(K,I)+V(K,I)*DT
50          CONTINUE
60        CONTINUE
C test for collisions
          DO 200 I=1,N
C test for collisions with wall
            DO 70 K=1,2
              IF(X(K,I).LE.XA(K)+R.OR.
     +        X(K,I).GE.XB(K)-R) V(K,I)=-V(K,I)
70          CONTINUE
C test for collisions with another particle
            IF(I.LT.N) THEN
              DO 100 J=I+1,N
                DX(1)=X(1,I)-X(1,J)
                DX(2)=X(2,I)-X(2,J)
                DD=DX(1)**2+DX(2)**2
                IF(DD.LE.DSQ) THEN
                  DINV=1.D0/SQRT(DD)
                  EPERP(1)=DX(1)*DINV
                  EPERP(2)=DX(2)*DINV
                  EPAR(1)=EPERP(2)
                  EPAR(2)=-EPERP(1)
                  CALL MTXGCL(V,VI,2,N,I)
                  CALL MTXGCL(V,VJ,2,N,J)
                  CALL MTXADV(VI,VJ,W,2)
                  CALL MTXMSV(W,W,0.5D0,2)
                  CALL MTXSBV(VI,W,UI,2)
                  CALL MTXDOT(UI,EPERP,UIPERP,2)
                  CALL MTXDOT(UI,EPAR,UIPAR,2)
                  UIPERP=-UIPERP
                  CALL MTXMSV(EPERP,H1,UIPERP,2)
                  CALL MTXMSV(EPAR,H2,UIPAR,2)
                  CALL MTXADV(H1,H2,UI,2)
                  CALL MTXADV(UI,W,VI,2)
                  CALL MTXMSV(UI,UJ,-1.D0,2)
                  CALL MTXADV(UJ,W,VJ,2)
                  CALL MTXPCL(V,VI,2,N,I)
                  CALL MTXPCL(V,VJ,2,N,J)
                END IF
100           CONTINUE
            END IF
200       CONTINUE
1000    CONTINUE
        END
